### PR TITLE
Refactor session config loading customization

### PR DESCRIPTION
**Pull request by stealthycoin**
Adds a new session component to give more fine grained control over config variable loading. This PR maintains backwards compatibility with the existing `session_instance_vars` system while adding a new provider based component to the session. 

All of the classes located in `botocore/configprovider.py` are intended to be a part of the public API of a session and will need to maintain backwards compatibility.

cc @kyleknap @jamesls @JordonPhillips @dstufft @joguSD

### CODE REVIEW

File: botocore/configprovider.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+            providers.append(
+                ClosedDictProvider(
+                    keys=env_vars,
+                    source=self._environ,
+                )
+            )
+        if config_property is not None:
+            providers.append(
+                LazyProvider(
+                    provider_function=lambda: ClosedDictProvider(
```
Comment (kyleknap) [inline botocore/configprovider.py]: It might make sense to rename `provider_function` to something like `provide_factory`. First reading this, I would have assumed we would have been providing a lazily called function (similar to `provide()`) that provides the value.

File: botocore/configprovider.py [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+        if env_vars is not None:
+            providers.append(
+                ClosedDictProvider(
+                    keys=env_vars,
+                    source=self._environ,
+                )
+            )
+        if config_property is not None:
+            providers.append(
+                LazyProvider(
```
Comment (kyleknap) [inline botocore/configprovider.py]: I wonder if it makes sense to have a generalized `LazyProvider`? It took me a while to follow the laziness with the use of `lambda`. I could also see us implement specialized providers such as `SessionInstanceProvider` or `ScopedConfigProvider` that just calls the slow method when `provide()` gets called. Another nice part about this approach is that we would be able to cache the responses back if we plan to call the slow method a lot and it is slow.

File: botocore/utils.py [hunk 3]
```diff
@@ -245,25 +240,106 @@ def set_value_from_jmespath(source, expression, value, is_first=True):
 
 class InstanceMetadataFetcher(object):
 
-    _REQUIRED_CREDENTIAL_FIELDS = [
-        'AccessKeyId', 'SecretAccessKey', 'Token', 'Expiration'
-    ]
+    class _RetriesExceededError(Exception):
```
Comment (kyleknap) [inline botocore/utils.py]: What's the reasoning for moving this error class to inside this class?

File: botocore/utils.py [hunk 4]
```diff
@@ -245,25 +240,106 @@ def set_value_from_jmespath(source, expression, value, is_first=True):
...
         self._disabled = env.get('AWS_EC2_METADATA_DISABLED', 'false').lower()
         self._disabled = self._disabled == 'true'
         self._user_agent = user_agent
         self._session = botocore.httpsession.URLLib3Session(
             timeout=self._timeout,
-            proxies=get_environ_proxies(self._url),
+            proxies=get_environ_proxies(self._base_url),
+        )
+
+    def get_request(self, url_path, custom_retry=None):
```
Comment (kyleknap) [inline botocore/utils.py]: I think for the signature here we just expose a `needs_retry` that overrides whatever the general `_needs_retry` method. It is a little weird that no matter what a user of this interface cannot override the default empty/non-200 retry logic.

File: botocore/session.py [hunk 5]
```diff
@@ -972,6 +892,52 @@ def lazy_register_component(self, name, no_arg_factory):
...
+    def __delitem__(self, key):
+        del self._store[self.__keytransform__(key)]
+
+    def __iter__(self):
+        return iter(self._store)
+
+    def __len__(self):
+        return len(self._store)
+
+    def __keytransform__(self, key):
```
Comment (kyleknap) [inline botocore/session.py]: Can you elaborate on why we need this method?

File: botocore/session.py [hunk 6]
```diff
@@ -972,6 +892,52 @@ def lazy_register_component(self, name, no_arg_factory):
             pass
 
 
+class SessionVarsShim(collections.MutableMapping):
```
Comment (kyleknap) [inline botocore/session.py]: I think it would be fine to just call this `SessionVarsMap`. The `Shim` part is a little vague for me.

File: botocore/session.py [hunk 7]
```diff
@@ -972,6 +892,52 @@ def lazy_register_component(self, name, no_arg_factory):
...
+    def __init__(self, session, *args, **kwargs):
+        self._session = session
+        self._store = dict()
+        self.update(dict(*args, **kwargs))
```
Comment (kyleknap) [inline botocore/session.py]: This is a little hard to follow what is being passed in here. Is there a reason we do not just accept the original `session_vars_map` dictionary as an optional argument? The nice part about that is we would no longer need an `if` statement when instantiating this.

File: botocore/configprovider.py [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+            default=None,
+        ),
+        'data_path': chain_builder.build_config_chain(
+            env_vars='AWS_DATA_PATH',
+            config_property='data_path',
+            default=None,
+        ),
+        'config_file': chain_builder.build_config_chain(
+            env_vars='AWS_CONFIG_FILE',
+            default='~/.aws/config',
```
[the remark this answers is not in the document]
Comment (stealthycoin) [reply botocore/configprovider.py]: Not sure, this is unchanged from before.

File: botocore/utils.py [hunk 9]
```diff
@@ -243,28 +238,108 @@ def set_value_from_jmespath(source, expression, value, is_first=True):
...
+    """Internal exception used when the number of retries are exceeded."""
+    pass
+
+
 class InstanceMetadataFetcher(object):
 
-    _REQUIRED_CREDENTIAL_FIELDS = [
-        'AccessKeyId', 'SecretAccessKey', 'Token', 'Expiration'
-    ]
+    _RETRIES_EXCEEDED_ERROR = _RetriesExceededError
```
Comment (kyleknap) [inline botocore/utils.py]: We should rename this to `_RETRIES_EXCEEDED_ERROR_CLS` and raise instances of the class. Right now I think we are just raising the exception class.

File: tests/unit/test_config_provider.py [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+        self.assertEqual(value, expected_value)
+
+    def test_chain_builder_can_provide_instance(self):
+        self.assert_chain_does_provide(
+            instance_map={'foo': 'bar'},
+            environ_map={},
+            config_map={},
+            build_config_chain_args={},
+            logical_name_to_provide='foo',
+            expected_value='bar',
```
Comment (kyleknap) [inline tests/unit/test_config_provider.py]: So for these tests to make it even more readable, it would be nice to put the name of where it came from as the value so instead of using `bar` make the value `instance-val` to indicate it came from the instance values. You did it a little bit later on for the keys, but it would be nice to do it for the values as well.

File: tests/unit/test_config_provider.py [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+            **build_config_chain_args
+        )
+        value = chain.provide()
+        self.assertEqual(value, expected_value)
+
+    def test_chain_builder_can_provide_instance(self):
+        self.assert_chain_does_provide(
+            instance_map={'foo': 'bar'},
+            environ_map={},
+            config_map={},
```
Comment (kyleknap) [inline tests/unit/test_config_provider.py]: Maybe call this the `scoped_config_map`? It is a little more specific on what config is being referenced

File: botocore/configprovider.py [hunk 12]
```diff
@@ -0,0 +1,416 @@
...
+        :type logical_name: str
+        :param logical_name: The logical name of the config value that this
+            chain is responsible for loading.
+
+        :type instance: bool
+        :param instance: This indicates whether or not session instance
+            variable lookup should be included in the config chain. By default
+            it is True.
+
+        :type env_vars: str or list of str or None
```
Comment (kyleknap) [inline botocore/configprovider.py]: Just a little more feedback on the names after looking at the tests... Maybe call this `env_varnames` or something else to indicate it is a name? Typically, `env_vars` or `env` is associated as a map and it threw me off when I was reading through the tests. The same thing goes for `config_property`, but not as much. Maybe try to find a parameter name that better indicates it expects the `name`?

File: tests/unit/test_config_provider.py [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+def assert_chain_does_provide(providers, expected_value):
+    provider = ChainProvider(
+        providers=providers,
+    )
+    value = provider.provide()
+    assert_equal(value, expected_value)
+
+
+def test_chain_provider():
+    # Each case is a tuple with the first element being the expected return
```
Comment (kyleknap) [inline tests/unit/test_config_provider.py]: Not too picky about this, but if you wanted you could copy a bit of what we do in [testing s3 addressing](https://github.com/boto/botocore/blob/7052b0f91fca61bf5206ec73240a99094fd2207a/tests/functional/test_s3.py#L855) where we have a class to specify cases. The nice part about this is that you can have parameter names so we would not have to explain the parameters in the comments and they would be self-documenting.

File: tests/unit/test_session.py [hunk 14]
```diff
@@ -97,49 +138,28 @@ def test_profile_when_pulled_from_env(self):
...
-        }
-        session = create_session(session_vars=env_vars)
+        self.update_session_config_mapping(
+            'profile', env_vars=['BAR_DEFAULT_PROFILE', 'BAR_PROFILE']
+        )
         self.environ.pop('BAR_DEFAULT_PROFILE', None)
         self.environ['BAR_PROFILE'] = 'second'
-        self.assertEqual(session.get_config_variable('profile'), 'second')
-
-    def test_profile(self):
```
Comment (kyleknap) [inline tests/unit/test_session.py]: Why did this test get removed?

Comment (stealthycoin) [reply tests/unit/test_session.py]: Wasn't doing anything useful.

File: tests/unit/test_config_provider.py (hunk 13 above)
Comment (stealthycoin) [reply tests/unit/test_config_provider.py]: Personally I like the way I did it better for this case since its super easy to see the chain and the result it should give with no extra verbosity.

File: botocore/configprovider.py [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+        ),
+        'parameter_validation': chain_builder.create_config_chain(
+            instance_name='parameter_validation',
+            config_property_name='parameter_validation',
+            default=True,
+        ),
+    }
+
+
+class ConfigChainFactory(object):
```
Comment (jamesls) [inline botocore/configprovider.py]: Would you consider this part of the public API?  Seems like we really only want the `ConfigValueStore` to be considered public right?  It's worth documenting what we consider to be public here.

File: botocore/credentials.py [hunk 16]
```diff
@@ -67,7 +67,7 @@ def create_credential_resolver(session, cache=None):
     env_provider = EnvProvider()
     container_provider = ContainerProvider()
     instance_metadata_provider = InstanceMetadataProvider(
-        iam_role_fetcher=InstanceMetadataFetcher(
+        iam_role_fetcher=InstanceMetadataCredentialFetcher(
```
Comment (jamesls) [inline botocore/credentials.py]: I'm a little concerned about this impacting people.  I'd prefer we keep the name the same, and just create something new like `IMDSClient` for making arbitrary requests to IMDS.  We can revisit the names in a major version bump.

File: botocore/utils.py [hunk 17]
```diff
@@ -243,28 +238,108 @@ def set_value_from_jmespath(source, expression, value, is_first=True):
...
         self._disabled = env.get('AWS_EC2_METADATA_DISABLED', 'false').lower()
         self._disabled = self._disabled == 'true'
         self._user_agent = user_agent
         self._session = botocore.httpsession.URLLib3Session(
             timeout=self._timeout,
-            proxies=get_environ_proxies(self._url),
+            proxies=get_environ_proxies(self._base_url),
         )
 
+    def get_request(self, url_path, retry_func=None):
```
Comment (jamesls) [inline botocore/utils.py]: So I didn't mind this as an internal method on the original imds fetcher class, but this just seems like a weird public interface to me (specifically the `retry_func` part).  Is there any way we can just keep the retry stuff in the original class that fetches iam role credentials?  This class could just return the response and the class that's retrieving the iam role could still decide if it needed to retry.

File: botocore/configprovider.py (hunk 15 above)
Comment (stealthycoin) [reply botocore/configprovider.py]: Ah I hadn't thought about that specifically. I don't think there is any reason for it to be public since its a convenience class for us. It would only be useful for people that want to make the same configuration chain as we do, which I don't think is a big use case. Mainly I was considering the `ConfigStore` as the public class since it gets exposed through the `get_component('config_store')` on the session. And thats the main way to interact with overriding config providers. From there the `BaseProvider` is the interface they would use so that is public as well.

Should we add this to the botocore docs somewhere or document it in code?

File: botocore/credentials.py (hunk 16 above)
Comment (stealthycoin) [reply botocore/credentials.py]: Ah good point, I can change the names around.

File: botocore/utils.py (hunk 17 above)
Comment (stealthycoin) [reply botocore/utils.py]: This was to make it easier to write more clients since they all share that functionality. Making it public was to indicate that its a publicly facing function since it will have a consumer in a different project (CLI).
